### PR TITLE
Iss53squash

### PR DESCRIPTION
**Pull request by hustf**
Note that no docbuild was attempted. "Exteral documentation" has a few names replaced.

See commit message https://github.com/kwehmeyer/Spotify.jl/commit/a47210660e523483c4e711e8abfbe3f734fa0d83

### CODE REVIEW

**Comment by hustf** [comment]
Fixes issue #53.

**Comment by hustf** [comment]
Trying to install on a new machine, I found an issue with writing the default .ini file - it seems to be in JSON format. Also, for those who have existing .ini-files, warnings ought to be added regaring adding 'user' and 'country' files. Will update this PR shortly.

**Comment by hustf** [comment]
Documentation build will still fail. It runs on Julia 1.6. 'Iterators.Pairs' was evidently moved to 'Base.Pairs'. I suggest using a newer version.

File: Project.toml
Comment (kwehmeyer) [inline Project.toml]: @hustf feel free to add yourself to this list :)

**Review by kwehmeyer — APPROVED** [review]

**Comment by kwehmeyer** [comment]
@hustf if you could add a summary of the PR to the description that would be great. Doesn't need to be super detailed.

> Added X for the purpose of Y

> Changed X to do Y

etc., 

Once thats in I think it's good to merge!

**Comment by hustf** [comment]
Thanks for the approval, Kwehmeyer! I'll be honored to put myself in the list!

Purpose? That's deep. I need to think about that and come back. Also, I'd appreciate your thoughts regarding what to do about the documentation? 

Meanwhile, in the web of commits and regrets comprising this PR, this [summary](https://github.com/kwehmeyer/Spotify.jl/commit/a47210660e523483c4e711e8abfbe3f734fa0d83) got stuck and buried. Here, I pulled it out:


## User permissions

`using Spotify` will now ask user for limited privileges through
clicking in a pop-up browser.

-["user-read-private", "user-modify-playback-state",
    "user-read-playback-state", "playlist-modify-private",
    "playlist-read-private"]

If user makes calls which need more, the list is extended and
new user permission is applied for. Grants expire after one
hour. User is warned to call `authorize()`.


## Log control, general requests.

- Introduce detail log control, hide non-error details as default.
- HTTP.jl blabber is silenced, which requires adding 'Logging'
from Base.
- For error codes returned, show both the general error
codes and their meaning. Also show detail feedback
which is included in some error message bodies. When errors
are logged (in red), the request is also shown (in red).
- Empty arguments (e.g. 'country=""&offset=0') are not sent. See 
  'Argument types' below.


## `Market`, `locale` and `country` arguments

Where such arguments are necessary, default values are taken from the .ini file, 
which has a new field for this. For other functions, the default behavior is not to 
send these arguments.


## Menus and maintainability

A rudimentary menu existed, but required maintenance. Menu
now in src, using introspection to build menu.Menu's
argument values are now based on names, not types. Output
is formatted for making inline docs examples.


## Argument types

Use duck-typing in all function wrappers, but internal conversion
to SpArtistId, etc. Sp___Id types have one representation
in url strings, and one including type in body strings. Export function 
`parse_copied_link', which finds type.


## Module structure

Move wrapper functions to mirror the 'Endpoints' hierarcy.
A few are only found in the 'Console' hierarchy. These get
alias functions.


## Inline docs

Consistent formatting. Will hopefully work with docbuild, but
the 'docs' folder is hardly touched and might require work.

## Examples folder, add files

- `playlist_and_library_utilties.jl` for recurring functions.
- `mini_player.jl` ( ':' to enter REPL mode).
- `remove_duplicates_in_same_playlists.jl`

**Comment by hustf** [comment]
# My purpose 

with this submission is to prepare for a public transport regularity analysis.
These parts are quite reusable:

- The limited type system, where representation depends on context: JSON, terminal, 
    or URI.
- One 'request' function for all endpoints, including authorization fields.
- Duck typing with internal conversion.

# Package's purpose 

in pulling this would be to improve usability to that stage where any Julia user 
can easily adapt examples to fit in ones own algorithms.

To this purpose, every call outputs the tuple '(Json3Object, Integer)' as before. Examples are 
one level up from the web API interface. This level is where user 
 - parses returned JSON3Object, perhaps to Spotify.jl's types.
 - stores in Dict, LightXML, DataFrame or maybe jute sacks? 

# Problem fixes

 - ❌ means 'has a problem' or 'not working'
- ✓  means 'fixed'
- ⌅ means 'limits to'
- => means 'leads to'


| Problems as experienced on Windows 11  |v0.1.2| v0.2 | Ref.                 |
| :---                                   |:---: | :---:| :---                 |
| test_playlists.jl test_player.jl       |  ❌  |   ✓  |                     |
| test_tracks.jl test_browse.jl          |  ❌  |   ✓  |                     |
| test_episodes.jl                       |  ❌  |   ✓  |                     |
| HTTP ⌅ 0.9.17 => ⌅Compat ⌅TerminalMenus|  ❌  |   ✓  | Project.toml        |
| Warnings: Could not export etc..       |  ❌  |   ✓  | export_structure.jl |
| Browser launch not working             |  ❌  |   ✓  |implicit_grant_flow.jl|
| Default is 'Client credentials' only   |  ❌  |   ✓  | 1) below            |
| Feedback from errored calls partly lost|  ❌  |   ✓  |util/request.jl -> src|
| Some function names may confuse        |  ❌  |   ✓  | 2)                  |
| Web docs have several hierarchies      |  ❌  |   ❌ | 3)                  |
| 'csv_of_saved_tracks.jl' (master only) |  ❌   |  ✓  | 4)                  |
| Terminal messages not targeted to user |  ❌   |  ✓  |?LOGSTATE, request.jl|
| Request body feature for web API calls |  ❌   |  ✓  | 5)                  |
| Some country / market args hardcoded   |  ❌   |  ✓  | .ini field, also drop args|


1)  'Client credentials' require no user grants and is quite limited. Default 
   rate limits are severe, but less so with a user grant. It will be hard
   to do useful analysis without implicit grants. Hence, we now immediately 
   `apply_and_wait_for_implicit_grant(;scopes =..)`

2) Function names now describe what they do in calling context:

    - `implicit_grant_flow/request_handler` -> `receive_grant_as_request`
    - `start_a_listening_server`  -> `launch_async_single_grant_receiving_server`


3) Considering: 

    - 'reference/endpoints' is the complete hierarchy.
    - The 'Console' tree contains the necessary grant scopes.
    
  File folders

    - 'by_reference_doc' folder now accurately reflects the former.
    - 'by_console_doc' contain useful alias functions. Deleted file 'follow.jl',
       whose functions now rests in 'users.jl'.

4) This (unreleased) example required running in a separate environment with 
   DataFrames and CSV. But HTTP ⌅ 0.9.17 also limited these packages to ancient
   versions.

   Renamed file to 'csv_of_liked_tracks.jl', added explaining comments, pruned bitrot.

5) Some calls can convey arguments through urlstring as an alternative to a body.
    Usually, urlstring has more severe limitations. Wrappers which now sends arguments as a body:
    - `users_unfollow_artists_users`
    - `users_follow_playlist`
    - `playlist_create_playlist`
    - `playlist_remove_playlist_item`
    - `playlist_add_tracks_to_playlist`
    - `playlist_change_details`
    - `player_resume_playback`

**Comment by hustf** [comment]
The latest commit added me to the authors list as recommended, and made minimal changes for a successful docbuild.
I believe this is ready to be merged, but I don't have write access to do so.

**Pull request merged**
